kernel/Kconfig

```diff
@@ -1083,6 +1083,13 @@ config DEVICE_DEPS_DYNAMIC
 	  Option that makes it possible to manipulate device dependencies at
 	  runtime.
 
+config DEVICE_MUTABLE
+	bool "Mutable devices [EXPERIMENTAL]"
+	select EXPERIMENTAL
+	help
+	  Support mutable devices. Mutable devices are instantiated in SRAM
+	  instead of Flash and are runtime modifiable in kernel mode.
```

Comment:
I think mutable devices should be documented clearly to avoid future misusages. This is a bit too generic IMHO.

Reply:
> I think mutable devices should be documented clearly to avoid future misusages.

Clearly documented where / how? Please feel free to suggest a specific preference.

> This is a bit too generic IMHO.

Hehe.. well, it's definitely possible to specialize things for more specialized use cases, but I think that might be getting into "premature specialization".

Maybe a DT property could be added allowing one to specify the initial selection. But at the end of the day, it's a multiplexer (switch) - there really isn't much more to the basic requirement.

Also, we have devicetree to aid in specialization through inheritance. So if specialization needs to happen for a particilar subclass, it might be better to add e.g. new properties as necessary, rather than forcing them in the base case.

Reply:
@cfriedt I'd really like to see a code sample (+README)  actually. For example could this be used to transparently expose a modem's AT shell as zephyr,shell-uart so that one can tinker with AT commands directly and then, with the push of a button, one can go back to the Zephyr shell? Or am I completely off?

Reply:
And if I'm not off then I guess the sample should also showcase the simpler version where one only changes "selected" in the devicetree. The button thing is just to show the dynamic aspect.

Reply:
Oh weird - my comment didn't seem to submit properly from the GH app. 

> one can go back to the Zephyr shell? Or am I completely off?

Yeah, that's basically the gist of it, although there are many applications. 

>  I'd really like to see a code sample (+README) actually. 

There is the existing test obviously that shows what's necessary to switch between multiplexed uart devices in software. Extrapolating that out to a sample app is pretty trivial.

>  the simpler version where one only changes "selected" in the devicetree.

Changing `selected` in Devicetree just allows for explicit specification of the initial selection (which is otherwise set to position zero). The devmux device is still fully functional and isn't necessarily any simpler.

I guess a sample app would help to clarify those things, so I will get to it 😁

Reply:
Did you post in the wrong comment thread? If yes could you move the comment above to the other PR for reference? FWIW my understanding is that we're not even adding new code samples during feature freeze so I am not sure I would agree with your statement.

Reply:
I thought you had asked for a sample. At first, I thought sure, but as you mentioned we are past feature freeze. I'd prefer not to add a sample here in this case (mainly because this change is really just being used for recovering unresponsive Zephyr FW).

Reply:
> I thought you had asked for a sample. At first, I thought sure, but as you mentioned we are past feature freeze. I'd prefer not to add a sample here in this case (mainly because this change is really just being used for recovering unresponsive Zephyr FW).

Oh yes I'm totally asking for a sample, sorry if this wasn't clear! But I guess the misunderstanding is that I, and I believe everyone else in this thread, is just discussing in the context of 3.6, to make sure this is ready to go as soon as the doors open again.

I don't see how this can make it for 3.5, unless this was discussed in today's release meeting which I missed? But this would also require a TSC vote IIUC.